wlroots/ffi_build.py

```diff
@@ -2237,6 +2251,245 @@ def check_version():
+        struct wlr_xwayland_surface *surface;
+        bool minimize;
+        ...;
+    };
+    struct wlr_xwayland_server *wlr_xwayland_server_create(
+        struct wl_display *display, struct wlr_xwayland_server_options *options);
+    void wlr_xwayland_server_destroy(struct wlr_xwayland_server *server);
+    struct wlr_xwayland *wlr_xwayland_create(struct wl_display *wl_display,
+        struct wlr_compositor *compositor, bool lazy);
+    void wlr_xwayland_destroy(struct wlr_xwayland *wlr_xwayland);
+    void wlr_xwayland_set_cursor(struct wlr_xwayland *wlr_xwayland,
+        uint8_t *pixels, uint32_t stride, uint32_t width, uint32_t height,
+        int32_t hotspot_x, int32_t hotspot_y);
+    void wlr_xwayland_surface_activate(struct wlr_xwayland_surface *surface,
+        bool activated);
+    void wlr_xwayland_surface_restack(struct wlr_xwayland_surface *surface,
+        struct wlr_xwayland_surface *sibling, enum xcb_stack_mode_t mode);
+    void wlr_xwayland_surface_configure(struct wlr_xwayland_surface *surface,
+        int16_t x, int16_t y, uint16_t width, uint16_t height);
+    void wlr_xwayland_surface_close(struct wlr_xwayland_surface *surface);
+    void wlr_xwayland_surface_set_minimized(struct wlr_xwayland_surface *surface,
+        bool minimized);
+    void wlr_xwayland_surface_set_maximized(struct wlr_xwayland_surface *surface,
+        bool maximized);
+    void wlr_xwayland_surface_set_fullscreen(struct wlr_xwayland_surface *surface,
+        bool fullscreen);
+    void wlr_xwayland_set_seat(struct wlr_xwayland *xwayland,
+        struct wlr_seat *seat);
+    bool wlr_surface_is_xwayland_surface(struct wlr_surface *surface);
+    struct wlr_xwayland_surface *wlr_xwayland_surface_from_wlr_surface(
+        struct wlr_surface *surface);
+    void wlr_xwayland_surface_ping(struct wlr_xwayland_surface *surface);
+    bool wlr_xwayland_or_surface_wants_focus(
+        const struct wlr_xwayland_surface *xsurface);
+    enum wlr_xwayland_icccm_input_model wlr_xwayland_icccm_input_model(
+        const struct wlr_xwayland_surface *xsurface);
+
+    typedef struct xcb_intern_atom_cookie_t {
+        ...;
+    } xcb_intern_atom_cookie_t;
+    typedef ... xcb_connection_t;
+    typedef struct xcb_intern_atom_reply_t {
+        uint8_t    response_type;
+        uint8_t    pad0;
+        uint16_t   sequence;
+        uint32_t   length;
+        xcb_atom_t atom;
+        ...;
+    } xcb_intern_atom_reply_t;
+    typedef ... xcb_generic_error_t;
+
+    int xcb_connection_has_error(xcb_connection_t *c);
+    xcb_connection_t *xcb_connect(const char *displayname, int *screenp);
+    void xcb_disconnect(xcb_connection_t *c);
+    xcb_intern_atom_cookie_t xcb_intern_atom(xcb_connection_t *conn, uint8_t only_if_exists,
+        uint16_t name_len, const char *name);
+    xcb_intern_atom_reply_t *xcb_intern_atom_reply(xcb_connection_t *conn,
+        xcb_intern_atom_cookie_t cookie, xcb_generic_error_t **e);
```

Comment:
It would be great if we could use xcffib here. I'm not sure how possible that is, since we have used the API mode for the cffi bindings in pywayland and pywlroots, where we were forced to use ABI mode for xcffib (to be able to use xcffib in cairocffi, which is also ABI mode). That said, if we only need a tiny subset of xcb functionality here, we wan just do that ourselves to avoid the dependency problem with xcffib.

Reply:
Yes my thinking here was to avoid the complexity and dependency of compiling with another ffi lib. These few functions seem to be the only ones needed, and only to fetch the values for some atoms that are stored on `wlr_xwayland`, such as supported protocols and wm_types, that kind of thing. This seems to be common across sway, dwl, and a couple other compositors I checked. I'm happy to go with whichever you choose though!

Reply:
That sounds good, I was curious what your thoughts were, and I think I agree with your thinking here!